File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-2.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        char[] cs = s.toCharArray();
        int len = 0;
        int left = 0;
        Map<Character, Integer> map = new HashMap<>();
        for (int right = 0; right < cs.length; right++) {
            map.put(cs[right], map.getOrDefault(cs[right], 0) + 1);
            while (map.get(cs[right]) > 1) {
                map.put(cs[left], map.get(cs[left]) - 1);
                left++;
            }
            len = Math.max(len, right - left + 1);
        }
        return len;
    }
}

File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-2.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        cnt = defaultdict(int)

        left = max_count = max_len = 0
        for right, c in enumerate(s):
            cnt[c] += 1
            max_count = max(max_count, cnt[c])
            if (right - left + 1) - max_count > k:
                cnt[s[left]] -= 1
                left += 1
            max_len = max(max_len, right - left + 1)
        return max_len